File: GapBuffer.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

char *buffer;
int capacity = 50;
int gap_left = 0;
int gap_right = 9;

void grow(int gap_extra_size){
    char *new_buffer = malloc(capacity + gap_extra_size);
    for (int i = 0; i < capacity + gap_extra_size; i++) {
		new_buffer[i] = '_';
	}
    
    memcpy(new_buffer, buffer, gap_left);
    int right_size = capacity - gap_right - 1;
    memcpy(
        new_buffer + gap_left + gap_extra_size,
        buffer + gap_right + 1,
        right_size
    );
    
    gap_right = gap_left + gap_extra_size - 1;
    free(buffer);
    buffer = new_buffer;
    capacity += gap_extra_size;
}

void move_left(int position){
    while(position < gap_left){
		gap_left--;
		gap_right--;
		buffer[gap_right + 1] = buffer[gap_left];
		buffer[gap_left] = '_';
	}
}

void move_right(int position){
   	while(position > gap_left){
		gap_left++;
		gap_right++;
		buffer[gap_left-1] = buffer[gap_right];
		buffer[gap_right] = '_';
	}
}

void move_cursor(int position){
    if (position < gap_left){
        move_left(position);
    }else {
        move_right(position);
    }
    
}

void remove_string(int position){
    if (position != gap_left) {
        move_cursor(position);
    }
    buffer[gap_right + 1] = '_';
}


void insert(const char *input, int position){
    int len = strlen(input);
    
    if (gap_right - gap_left + 1 < len){
        grow(len > 10 ? len : 10);
    }
    
    if (position != gap_left) {
        move_cursor(position);
    }
    
    for (int i = 0; i < len; i++){
        buffer[gap_left++] = input[i];
    }
}

void print_buffer(){
    for (int i = 0; i < capacity; i++) {
        putchar(buffer[i]);
    }
    putchar('\n');
}


int main(){
    buffer = malloc(capacity);
    for (int i = 0; i < capacity; i++) {
        buffer[i] = '_';
    }
    
    insert("Hello ld", 0);
    print_buffer();

    insert("ta", 2);
    print_buffer();

    printf("l : %d\n", gap_left);
    printf("r : %d\n", gap_right);

    insert("bu", 20);
    print_buffer();

    printf("l : %d\n", gap_left);
    printf("r : %d\n", gap_right);

    remove_string(2);
    print_buffer();

    free(buffer);
    return 0;
    
}

File: GapBuffer.java
public class GapBuffer {
	private static char[] buffer =  new char[50];
	private static int gap_left = 0;
	private static int gap_right = 9;

	static void grow(int extra_gap_size) {
		char[] new_buffer = new char[buffer.length + extra_gap_size];

		for (int i = 0; i < new_buffer.length; i++) {
			new_buffer[i] = '_';
		}

		System.arraycopy(
				buffer,
				0,
				new_buffer,
				0,
				gap_left
		);

		int right_side = buffer.length - gap_right - 1;
		System.arraycopy(
				buffer,
				gap_right + 1,
				new_buffer,
				gap_left + extra_gap_size,
				right_side
		);

		gap_right = gap_left + extra_gap_size - 1;
		buffer = new_buffer;
	}


	static void move_cursor(int position){
		if (position < gap_left) {
			left(position);
		}else {
			right(position);
		}
	}

	static void left(int position){
		while(position < gap_left){
			gap_left--;
			gap_right--;
			buffer[gap_right + 1] = buffer[gap_left];
			buffer[gap_left] = '_';
		}
	}

	static void right(int position){
		while(position > gap_left){
			gap_left++;
			gap_right++;
			buffer[gap_left-1] = buffer[gap_right];
			buffer[gap_right] = '_';
		}
	}

	static void printBuffer(){
		for (int i = 0; i < buffer.length;  i++) {
			System.out.print(buffer[i]);
		}
		System.out.println();
	}



	static void insert(String input, int position){
		int length = input.length();

		if (gap_right - gap_left + 1 < length) {
			grow(Math.max(10, length));
		}

		if (position != gap_left){
			move_cursor(position);
		}
		System.out.println("left : " + gap_left);
		System.out.println("right : " + gap_right);

		for (int i = 0; i < length; i++) {
			buffer[gap_left] =  input.charAt(i);
			gap_left++;
			position++;
		}
	}

	static void remove(int position){
		if (position != gap_left){
			move_cursor(position);
		}
		buffer[gap_right + 1] =  '_';

	}

	public static void main(String[] args) {

		for (int i = 0; i < buffer.length; i++){
			buffer[i] = '_';
		}

		String a = "Hello ld";
		insert (a,0);
	printBuffer();

		String b = "ta";
		insert(b, 2);
printBuffer();

		System.out.println("l : " + gap_left);
		System.out.println("r : " + gap_right);

		String c = "bu";
		insert (c, 20);
		printBuffer();
		System.out.println("l : " + gap_left);
		System.out.println("r : " + gap_right);

		remove(2);
		printBuffer();

	}
}